File: apollo/dataService.js
const humans = [
  {
    id: '1000',
    name: 'Luke Skywalker',
    friends: ['1002', '1003', '2000', '2001'],
    appearsIn: ['NEWHOPE', 'EMPIRE', 'JEDI'],
    height: 1.72,
    mass: 77,
    starships: ['3001', '3003'],
  },
  {
    id: '1001',
    name: 'Darth Vader',
    friends: ['1004'],
    appearsIn: ['NEWHOPE', 'EMPIRE', 'JEDI'],
    height: 2.02,
    mass: 136,
    starships: ['3002'],
  },
  {
    id: '1002',
    name: 'Han Solo',
    friends: ['1000', '1003', '2001'],
    appearsIn: ['NEWHOPE', 'EMPIRE', 'JEDI'],
    height: 1.8,
    mass: 80,
    starships: ['3000', '3003'],
  },
  {
    id: '1003',
    name: 'Leia Organa',
    friends: ['1000', '1002', '2000', '2001'],
    appearsIn: ['NEWHOPE', 'EMPIRE', 'JEDI'],
    height: 1.5,
    mass: 49,
    starships: [],
  },
  {
    id: '1004',
    name: 'Wilhuff Tarkin',
    friends: ['1001'],
    appearsIn: ['NEWHOPE'],
    height: 1.8,
    mass: null,
    starships: [],
  },
];

const humanData = {};
humans.forEach((human) => {
  humanData[human.id] = human;
});

const droids = [
  {
    id: '2000',
    name: 'C-3PO',
    friends: ['1000', '1002', '1003', '2001'],
    appearsIn: ['NEWHOPE', 'EMPIRE', 'JEDI'],
    primaryFunction: 'Protocol',
  },
  {
    id: '2001',
    name: 'R2-D2',
    friends: ['1000', '1002', '1003'],
    appearsIn: ['NEWHOPE', 'EMPIRE', 'JEDI'],
    primaryFunction: 'Astromech',
  },
];

const droidData = {};
droids.forEach((droid) => {
  droidData[droid.id] = droid;
});

const starships = [
  {
    id: '3000',
    name: 'Millenium Falcon',
    length: 34.37,
  },
  {
    id: '3001',
    name: 'X-Wing',
    length: 12.5,
  },
  {
    id: '3002',
    name: 'TIE Advanced x1',
    length: 9.2,
  },
  {
    id: '3003',
    name: 'Imperial shuttle',
    length: 20,
  },
];

const starshipData = {};
starships.forEach((ship) => {
  starshipData[ship.id] = ship;
});

const reviewData = {};
reviewData["NEWHOPE"] = [];
reviewData["EMPIRE"] = [];
reviewData["JEDI"] = [];

function getCharacter(id) {
  console.log(`getCharacter, id:${id}`);
  return humanData[id] || droidData[id];
}

function getHero(episode) {
  if (episode === 'EMPIRE') {
    return humanData['1000'];
  }
  return droidData['2001'];
}

function getHuman(id) {
  console.log(`getHuman, id:${id}`);
  return humanData[id];
}

function getDroid(id) {
  console.log(`getDroid, id:${id}`);
  return droidData[id];
}

function getFriends(id) {
  console.log(`getFriends, id:${id}`);
  const characters = { ...humanData, ...droidData };
  const character = characters[id];
  if (character) {
    const { friends } = character;
    return friends.map(fid => characters[fid]);
  }
  return null;
}

function getStarship(id) {
  console.log(`getStarship id=${id}`);
  return starshipData[id];
}

function getStarshipsByHumanId(id) {
  console.log(`getStarshipsByHumanId, HumanId=${id}`);
  const human = humanData[id];
  if (human) {
    const { starships } = human;
    return starships.map(sid => starshipData[sid]);
  }
  return null;
}

function search(text) {
  console.log(`search, text:${text}`);
  const re = new RegExp(text, 'i');
  const allData = [...humans, ...droids, ...starships];
  return allData.filter(obj => re.test(obj.name));
}

function createReview(episode, review) {
  console.log(`createReview episode:${episode}, commentary:${review.commentary}`);
  let reviews = reviewData[episode];
  if (reviews) {
    reviews.push(review);
  }
  return review;
}

function getReviews(episode) {
  console.log(`getReviews episode:${episode}`);
  return reviewData[episode];
}

function getDataType(id) {
  if (humanData[id]) {
    return 'Human';
  }
  if (droidData[id]) {
    return 'Droid';
  }
  if (starshipData[id]) {
    return 'Starship';
  }
  return null;
}

module.exports = {
  getCharacter,
  getHuman,
  getDroid,
  getHero,
  getDataType,
  getHero,
  getStarship,
  getFriends,
  getStarshipsByHumanId,
  search,
  createReview,
  getReviews
};

File: vertx/src/main/java/com/example/vertx/graphql/data/DataService.java
package com.example.vertx.graphql.data;

import io.reactivex.Single;
import lombok.extern.slf4j.Slf4j;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static com.example.vertx.graphql.data.Episode.*;
import static java.util.Arrays.asList;
import static java.util.Collections.emptyList;

/**
 * @author Gary Cheng
 */
@Slf4j
public class DataService {
    private String aboutMessage = "Star Wars GraphQL api v1.0";
    private Map<String, Human> humanData;
    private Map<String, Droid> droidData;
    private Map<String, Starship> starshipData;
    private Map<String, MovieCharacter> characterData;
    private Map<Episode, List<Review>> reviewData;

    public DataService() {
        this.initData();
    }

    /**
     * Find character by id
     *
     * @param id character id
     * @return
     */
    public Single<MovieCharacter> getCharacter(String id) {
        log.debug("getCharacter, id={}", id);
        return Single.just(characterData.get(id));
    }

    /**
     * Find human by id
     *
     * @param id human id
     * @return
     */
    public Single<Human> getHuman(String id) {
        log.debug("getHuman, id={}", id);
        return Single.just(humanData.get(id));
    }

    /**
     * Find Droid by id
     *
     * @param id Droid id
     * @return
     */
    public Single<Droid> getDroid(String id) {
        log.debug("getDroid, id={}", id);
        return Single.just(droidData.get(id));
    }

    /**
     * Find Starship by id
     *
     * @param id starship id
     * @return
     */
    public Single<Starship> getStarship(String id) {
        log.debug("getStarship, id={}", id);
        return Single.just(starshipData.get(id));
    }

    /**
     * Return Hero character by episode
     *
     * @param episode the episode
     * @return
     */
    public Single<MovieCharacter> getHero(Episode episode) {
        log.debug("getHero, episode={}", episode);
        return Single.just(characterData.get(episode == EMPIRE ? "1000" : "2001"));
    }

    /**
     * Find friends by character id
     *
     * @param id
     * @return
     */
    public Single<List<MovieCharacter>> getFriends(String id) {
        log.debug("getFriends, id={}", id);
        return Single.just(this.characterData.get(id).getFriends().stream()
            .map(cid -> characterData.get(cid))
            .collect(Collectors.toList()));
    }

    /**
     * Return starships by human id
     *
     * @param id
     * @return
     */
    public Single<List<Starship>> getStarshipsByHumanId(String id) {
        log.debug("getStarshipsByHumanId, id={}", id);
        return Single.just(this.humanData.get(id).getStarships().stream()
            .map(sid -> this.starshipData.get(sid))
            .collect(Collectors.toList()));
    }

    /**
     * Search by text
     *
     * @param text
     * @return
     */
    public Single<List<Object>> search(String text) {
        log.debug("search by text, text={}", text);
        List<Object> results = new ArrayList<>();
        results.addAll(characterData.values().stream().filter(c -> textMatch(c.getName(), text)).collect(Collectors.toList()));
        results.addAll(starshipData.values().stream().filter(s -> textMatch(s.getName(), text)).collect(Collectors.toList()));
        return Single.just(results);
    }

    /**
     * Add Review
     *
     * @param episode
     * @param review
     * @return
     */
    public Single<Review> addReview(Episode episode, Review review) {
        log.debug("addReview for episode {}", episode);
        reviewData.get(episode).add(review);
        return Single.just(review);
    }

    /**
     * Get Reviews by episode
     *
     * @param episode
     * @return
     */
    public Single<List<Review>> getReviews(Episode episode) {
        log.debug("getReviews for episode {}", episode);
        return Single.just(reviewData.get(episode));
    }

    /**
     * Return about message
     *
     * @return
     */
    public Single<String> getAboutMessage() {
        log.debug("getAboutMessage");
        return Single.just(this.aboutMessage);
    }

    /**
     * Set about message
     *
     * @param message
     */
    public Single<String> setAboutMessage(String message) {
        log.debug("setAboutMessage message={}", message);
        this.aboutMessage = message;
        return Single.just(this.aboutMessage);
    }

    private boolean textMatch(String source, String text) {
        return source.toUpperCase().indexOf(text.toUpperCase()) >= 0;
    }

    private void initData() {
        log.debug("initData");
        this.humanData = new HashMap<>();
        this.droidData = new HashMap<>();
        this.starshipData = new HashMap<>();
        this.characterData = new HashMap<>();
        this.addHuman(new Human(
            "1000",
            "Luke Skywalker",
            1.72,
            77.0,
            asList("1002", "1003", "2000", "2001"),
            asList(NEWHOPE, EMPIRE, JEDI),
            asList("3001", "3003"))
        );
        this.addHuman(new Human(
            "1001",
            "Darth Vader",
            2.02,
            136.0,
            asList("1004"),
            asList(NEWHOPE, EMPIRE, JEDI),
            asList("3002"))
        );
        this.addHuman(new Human(
            "1002",
            "Han Solo",
            1.8,
            80.0,
            asList("1000", "1003", "2001"),
            asList(NEWHOPE, EMPIRE, JEDI),
            asList("3000", "3003"))
        );
        this.addHuman(new Human(
            "1003",
            "Leia Organa",
            1.5,
            49.0,
            asList("1000", "1002", "2000", "2001"),
            asList(NEWHOPE, EMPIRE, JEDI),
            emptyList())
        );
        this.addHuman(new Human(
            "1004",
            "Wilhuff Tarkin",
            2.02,
            136.0,
            asList("1001"),
            asList(NEWHOPE),
            emptyList())
        );

        this.addDroid(new Droid(
            "2000",
            "C-3PO",
            asList("1000", "1002", "1003", "2001"),
            asList(NEWHOPE, EMPIRE, JEDI),
            "Protocol"
        ));
        this.addDroid(new Droid(
            "2001",
            "R2-D2",
            asList("1000", "1002", "1003"),
            asList(NEWHOPE, EMPIRE, JEDI),
            "Astromech"
        ));

        this.addStarship(new Starship(
            "3000",
            "Millenium Falcon",
            34.37
        ));
        this.addStarship(new Starship(
            "3001",
            "X-Wing",
            12.5
        ));
        this.addStarship(new Starship(
            "3002",
            "TIE Advanced x1",
            9.2
        ));
        this.addStarship(new Starship(
            "3003",
            "Imperial shuttle",
            20.0
        ));
        reviewData = new HashMap<>();
        reviewData.put(NEWHOPE, new ArrayList<>());
        reviewData.put(EMPIRE, new ArrayList<>());
        reviewData.put(JEDI, new ArrayList<>());
    }

    private void addHuman(Human human) {
        this.humanData.put(human.getId(), human);
        this.characterData.put(human.getId(), human);
    }

    private void addDroid(Droid droid) {
        this.droidData.put(droid.getId(), droid);
        this.characterData.put(droid.getId(), droid);
    }

    private void addStarship(Starship starship) {
        this.starshipData.put(starship.getId(), starship);
    }
}
